SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LegacyLoadBalancingStrategy.cpp

```diff
@@ -0,0 +1,212 @@
+						if (Data.component_id == SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID)
+						{
+							Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+							Translator.ApplyVirtualWorkerManagerData(ComponentObject);
+							bTranslatorIsReady = true;
+							for (uint32 VirtualWorker = 0; VirtualWorker < ExpectedWorkers; ++VirtualWorker)
+							{
+								if (Translator.GetServerWorkerEntityForVirtualWorker(VirtualWorker + 1)
+									== SpatialConstants::INVALID_ENTITY_ID)
+								{
+									bTranslatorIsReady = false;
+								}
+							}
+						}
+					}
+				}
+			}
+			WorkerTranslationRequest.Reset();
+		}
+		}
+	}
+}
+
+void FLegacyLoadBalancing::QueryTranslation(ISpatialOSWorker& Connection)
+{
+	if (WorkerTranslationRequest.IsSet())
+	{
+		return;
+	}
+
+	// Build a constraint for the Virtual Worker Translation.
+	Worker_ComponentConstraint TranslationComponentConstraint;
+	TranslationComponentConstraint.component_id = SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID;
+
+	Worker_Constraint TranslationConstraint;
+	TranslationConstraint.constraint_type = WORKER_CONSTRAINT_TYPE_COMPONENT;
+	TranslationConstraint.constraint.component_constraint = TranslationComponentConstraint;
+
+	Worker_EntityQuery TranslationQuery{};
+	TranslationQuery.constraint = TranslationConstraint;
+
+	WorkerTranslationRequest = Connection.SendEntityQueryRequest(EntityQuery(TranslationQuery), RETRY_UNTIL_COMPLETE);
+}
+
+void FLegacyLoadBalancing::Flush(ISpatialOSWorker& Connection)
+{
+	if (ConnectedWorkers.Num() == ExpectedWorkers && !bTranslatorIsReady)
+	{
+		QueryTranslation(Connection);
+	}
+}
+
+void FLegacyLoadBalancing::Init(TArray<FLBDataStorage*>& OutLoadBalancingData)
+{
+	if (PositionStorage)
+	{
+		OutLoadBalancingData.Add(PositionStorage.Get());
+	}
+	if (GroupStorage)
+	{
+		OutLoadBalancingData.Add(GroupStorage.Get());
+	}
+	if (AssignmentStorage)
+	{
+		OutLoadBalancingData.Add(AssignmentStorage.Get());
+	}
+}
+
+void FLegacyLoadBalancing::OnWorkersConnected(TArrayView<FLBWorkerHandle> InConnectedWorkers)
+{
+	for (auto& Worker : InConnectedWorkers)
+	{
+		ConnectedWorkers.Add(Worker);
+	}
+}
+
+void FLegacyLoadBalancing::OnWorkersDisconnected(TArrayView<FLBWorkerHandle> DisconnectedWorkers)
+{
+	// Not handled
+}
+
+void FLegacyLoadBalancing::TickPartitions(FPartitionManager& PartitionMgr)
+{
+	if (bCreatedPartitions)
+	{
+		return;
+	}
+	if (ConnectedWorkers.Num() == ExpectedWorkers && bTranslatorIsReady)
+	{
+		TMap<Worker_EntityId_Key, FLBWorkerHandle> WorkersMap;
+		for (auto Worker : ConnectedWorkers)
+		{
+			WorkersMap.Add(PartitionMgr.GetServerWorkerEntityIdForWorker(Worker), Worker);
+		}
+
+		for (VirtualWorkerId i = 1; i <= ExpectedWorkers; ++i)
+		{
+			Worker_EntityId ServerWorkerEntityId = Translator.GetServerWorkerEntityForVirtualWorker(i);
+			VirtualWorkerIdToHandle[i - 1] = WorkersMap.FindChecked(ServerWorkerEntityId);
+		}
+
+		if (Calculator != nullptr)
```

Comment:
Why support Calculator being nullptr?

Reply:
That's the "direct assignment" mode, some compatibility mode where the distributed load balancer still makes the computation, but the StrategyWorker has auth over the delegation component.
I'm not 100% clear on what that gives us, but Sami could tell you more.

Reply:
That should just be a specific load balancing data writer thing. Although reading atm.